modules/dbfs/couchbase.go

```diff
@@ -113,63 +124,97 @@ func (di *DatabaseImpl) CBGetFileVersion(fileID int64) (int64, error) {
 	}
 
 	var version int64
-	frag.Content("version", &version)
-	return version, err
-}
-
-// CBGetFileChanges returns the array of file changes for the given fileID
-func (di *DatabaseImpl) CBGetFileChanges(fileID int64) ([]string, error) {
-	cb, err := di.openCouchBase()
-	if err != nil {
-		return []string{}, err
-	}
-
-	frag, err := cb.bucket.LookupIn(strconv.FormatInt(fileID, 10)).Get("changes").Execute()
+	err = frag.Content("version", &version)
 	if err != nil {
-		return []string{}, err
+		return -1, ErrResourceNotFound
 	}
 
-	var changes []string
-	frag.Content("changes", &changes)
-
-	return changes, err
+	return version, err
 }
 
 // CBAppendFileChange mutates the file document with the new change and sets the new version number
-func (di *DatabaseImpl) CBAppendFileChange(fileID int64, baseVersion int64, changes []string) (int64, error) {
+// Returns the new version number, the missing patches, and an error, if any.
+func (di *DatabaseImpl) CBAppendFileChange(fileID int64, changes, prevChanges []string) (int64, []string, error) {
 	// TODO (non-immediate/required): verify changes are valid changes
 	cb, err := di.openCouchBase()
 	if err != nil {
-		return -1, err
+		return -1, nil, err
 	}
+	key := strconv.FormatInt(fileID, 10)
 
 	// optimistic locking operation
 	// check the version is accurate and get the object's cas,
 	// then use it in the MutateIn call to verify the document hasn't updated underneath us
-	frag, err := cb.bucket.LookupIn(strconv.FormatInt(fileID, 10)).Get("version").Execute()
+	frag, err := cb.bucket.LookupIn(key).Get("version").Get("usetemp").Execute()
 	if err != nil {
-		return -1, ErrResourceNotFound
+		return -1, nil, ErrResourceNotFound
 	}
 
 	cas := frag.Cas()
 	var version int64
-	frag.Content("version", &version)
+	var useTemp bool
+	err = frag.Content("version", &version)
+	if err != nil {
+		return -1, nil, ErrNoData
+	}
+	err = frag.Content("usetemp", &useTemp)
+	if err != nil {
+		return -1, nil, ErrNoData
+	}
+
+	minVersion := int64(math.MaxInt64)
+	transformedChanges := make([]string, len(changes))
+
+	// Build patch, transform changes against newer changes.
+	for i, changeStr := range changes {
+		change, err := patching.NewPatchFromString(changeStr)
+		if err != nil {
+			return -1, nil, errors.New("Failed to parse patch")
+		}
+
+		// check to make sure the patch is being applied to the most recent revision
+		if change.BaseVersion > version {
+			return -1, nil, ErrVersionOutOfDate
+		}
+
+		if minVersion > change.BaseVersion {
+			minVersion = change.BaseVersion
+		}
+		// For every patch, calculate the patches that it does not have.
+		startIndex := len(prevChanges) - int(version-change.BaseVersion)
```

Comment:
once we calculate this for one of the patches we're getting in the set, don't we know it for all of them?